### PR TITLE
F9Aup

### PR DESCRIPTION
**Pull request by Kulkarni-Kaustubh**
Signed-off-by: Kaustubh Kulkarni <kaustubh.j.kulkarni@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [x] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide
[collections_log.txt](https://github.com/Seagate/cortx-test/files/8314530/collections_log.txt)
[Log file.txt](https://github.com/Seagate/cortx-test/files/8314531/Log.file.txt)

### CODE REVIEW

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    def perform_upgrade(self, upgrade_image_version: str, exc: bool = True):
+        """Function calls upgrade and put return in queue object."""
+        LOGGER.info("Calling upgrade.")
+        resp = self.deploy_lc_obj.upgrade_software(self.master_node_obj, upgrade_image_version,
+                                                   self.prov_deploy_cfg["git_remote_path"], exc=exc)
+        self.que.put(resp)
+
+    @pytest.mark.run(order=1)
+    @pytest.mark.lc
+    @pytest.mark.comp_prov
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:106]: Please dont use comp_prov markers for cft tests

File: libs/prov/prov_k8s_cortx_deploy.py [hunk 2]
```diff
@@ -1503,3 +1503,15 @@ def update_sol_with_image(file_path: str, image_dict: dict) -> tuple:
...
+    def service_upgrade_software(node_obj, upgrade_image_version: str) -> tuple:
+        """
+        Helper function to upgrade.
+        :param node_obj: Master node(Logical Node object)
+        :param upgrade_image_version: Version Image to Upgrade.
+        :return: True/False
+        """
+        LOGGER.info("Upgrading CORTX image to version: %s.", upgrade_image_version)
+        upgrade_cmd = PROV_CFG['k8s_cortx_deploy']["upgrade_cluster"].format(upgrade_image_version)
+        cmd = "cd {}; {}".format(PROV_CFG['k8s_cortx_deploy']["git_remote_path"], upgrade_cmd)
```
Comment (priyankaborawake) [inline libs/prov/prov_k8s_cortx_deploy.py:1517]: Please add this in try except block and return false if failure in upgrade command

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        if self.collect_sb:
+            path = os.path.join(LOG_DIR, LATEST_LOG_FOLDER)
+            support_bundle_utils.collect_support_bundle_k8s(local_dir_path=path,
+                                                            scripts_path=
+                                                            self.deploy_conf['k8s_dir'])
+        resp = self.deploy_lc_obj.destroy_setup(self.master_node_list[0],
+                                                self.worker_node_list)
+        assert_utils.assert_true(resp)
+
+    def perform_upgrade(self, upgrade_image_version: str, exc: bool = True):
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:97]: This method is not required, service_upgrade_software can be used directly in test.

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+"""Provisioner Component level test cases for K8s CORTX Software Upgrade."""
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:21]: Please update the correct description

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        LOGGER.info("Current version: %s", installed_version)
+        LOGGER.info("Step 2: Done.")
+        LOGGER.info("Step 3: Check if installing version is higher than installed version.")
+        # TODO : Better way to compare two versions.
+        LOGGER.info("upgrade_image: %s", self.upgrade_image)
+        upgrade_image_version = self.upgrade_image.split(":")[1]
+        upgrade_version = upgrade_image_version.split("-")[1]
+        LOGGER.info("Installing CORTX image version: %s", upgrade_version)
+        if int(upgrade_version) <= int(installed_version):
+            assert_utils.assert_true(resp[0],"Installed version is same or higher than installing version." )
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:128]: Add False instead of resp[0]

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        else:
+            LOGGER.info("Installed version is lower than installing version.")
+        # TODO : Use data-pod when cortx services framework is done.
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:131]: Remove TODO as its already done.

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        resp = self.deploy_lc_obj.check_s3_status(self.master_node_obj, pod_prefix=cons.POD_NAME_PREFIX)
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Step 5: Start upgrade.")
+        resp = self.deploy_lc_obj.service_upgrade_software(self.master_node_obj, self.upgrade_image)
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Step 6: Check if installed version is equals to installing version.")
+        resp = HAK8s.get_config_value(self.master_node_obj)
+        assert_utils.assert_true(resp[0], resp[1])
+        new_installed_version = resp[1]['cortx']['common']['release']['version']
+        LOGGER.info("New CORTX image version: %s", new_installed_version)
```
Comment (priyankaborawake) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:142]: Please compare if  new_installed_version and upgrade_image_version  are equal.

**Review by priyankaborawake — CHANGES_REQUESTED** [review]
@Kulkarni-Kaustubh  added few comments, please check

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+            if CMN_CFG["nodes"][node]["node_type"].lower() == "master":
+                cls.master_node_obj = node_obj
+                cls.master_node_list.append(node_obj)
+            else:
+                cls.worker_node_list.append(node_obj)
+        resp = cls.deploy_lc_obj.check_s3_status(cls.master_node_obj, pod_prefix=cons.POD_NAME_PREFIX)
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Done: Setup operations finished.")
+
+    @classmethod
```
Comment (pragamj) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:79]: why classmethod

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+
+class TestK8CortxUpgrade:
+    """ This class contains test cases for K8s CORTX Software Upgrade. """
+
+    @classmethod
+    def setup_class(cls):
+        """Setup class"""
+        LOGGER.info("STARTED: Setup Module operations")
+        cls.que = queue.queue()
```
Comment (pragamj) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:49]: is this used ?

**Review by pragamj — APPROVED** [review]

**Review by priyankaborawake — APPROVED** [review]

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        """
+        if self.collect_sb:
+            path = os.path.join(LOG_DIR, LATEST_LOG_FOLDER)
+            support_bundle_utils.collect_support_bundle_k8s(local_dir_path=path,
+                                                            scripts_path=
+                                                            self.deploy_conf['k8s_dir'])
+
+
+    @pytest.mark.lc
+    @pytest.mark.tags("TEST-33660")
```
Comment (swapnil-khandare) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:89]: you may need to add marker for test

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        assert_utils.assert_true(resp[0], resp[1])
+        pod_name = self.master_node_obj.get_pod_name(pod_prefix=cons.POD_NAME_PREFIX)
+        assert_utils.assert_true(pod_name[0], pod_name[1])
+        resp = self.deploy_lc_obj.get_hctl_status(self.master_node_obj, pod_name[1])
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Cluster is up and all services are started.")
+        LOGGER.info("Test Completed.")
+
+    @pytest.mark.lc
+    @pytest.mark.tags("TEST-33669")
```
Comment (swapnil-khandare) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:141]: you may need to add marker for test

**Review by swapnil-khandare — APPROVED** [review]

**Review by seagate-sarang-sawant — APPROVED** [review]

File: config/prov/prov_test.yaml [hunk 12]
```diff
@@ -138,7 +138,7 @@ k8s_cortx_deploy:
   git_k8_repo: "https://github.com/Seagate/cortx-k8s.git"
   git_k8_repo_file: "https://raw.githubusercontent.com/Seagate/cortx-k8s/{}/k8_cortx_cloud/solution.yaml"
   exe_prereq: "sh prereq-deploy-cortx-cloud.sh"
-  k8s_dir: "/root/deploy-scripts/k8_cortx_cloud/"
```
Comment (pragamj) [inline config/prov/prov_test.yaml:141]: keep deploy-scripts

File: tests/prov/test_prov_k8s_cortx_upgrade.py [hunk 13]
```diff
@@ -23,6 +23,10 @@
 import logging
 import os
 import time
+<<<<<<< HEAD
+=======
```
Comment (pragamj) [inline tests/prov/test_prov_k8s_cortx_upgrade.py:27]: remove these <<<<<

**Pull request merged**